src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -55,6 +55,24 @@ public static Task WriteMessageAsync(this PipeWriter pipeWriter, byte[] messageD
                 return Task.FromException(new RpcException(SendingMessageExceedsLimitStatus));
             }
 
+            if (!serverCallContext.HttpContext.Response.HasStarted)
+            {
+                return pipeWriter.WriteMessageCoreInitialAsync(messageData, serverCallContext, flush);
+            }
+
+            return pipeWriter.WriteMessageCoreAsync(messageData, serverCallContext, flush);
+        }
+
+        private static async Task WriteMessageCoreInitialAsync(this PipeWriter pipeWriter, byte[] messageData, HttpContextServerCallContext serverCallContext, bool flush)
+        {
+            // Must call StartAsync before the first pipeWriter.GetSpan() in WriteHeader
+            await serverCallContext.HttpContext.Response.StartAsync();
```

Comment:
This is a new requirement in preview3 in order to use PipeWriter.GetSpan/GetMemory.

Reply:
This should be optimized to avoid the state machine in the common case of completing synchronously

Reply:
That's what the HasStarted check is for. IN the common case, we call WriteMessageCoreAsync instead of WriteMessageCoreInitialAsync

Reply:
Except, that's not the common case. The common case is that HasStarted isn't true and StartAsync completes synchronously. See https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http.Abstractions/src/Extensions/HttpResponseWritingExtensions.cs#L64 for a more complicated example

Reply:
Ah StartAsync can complete synchronously.